apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/AbstractR4ResourceProvider.java

```diff
@@ -449,24 +485,40 @@ Bundle createBundleFor(
       DateRangeParam serviceDate,
       OffsetLinkBuilder paging,
       BundleOptions bundleOptions) {
-    List<T> resources = new ArrayList<>();
-
-    for (ResourceTypeV2<T, ?> type : resourceTypes) {
-      List<?> entities;
-
-      entities =
-          claimDao.findAllByMbiAttribute(
-              type, mbi, bundleOptions.isHashed, lastUpdated, serviceDate);
-
-      resources.addAll(
-          entities.stream()
-              .filter(e -> !bundleOptions.excludeSamhsa || samhsaMatcher.hasNoSamhsaData(e))
-              .map(e -> transformEntity(type, e, bundleOptions.includeTaxNumbers))
-              .collect(Collectors.toList()));
-    }
-
-    // Enforces a specific sorting for pagination that parities the EOB resource sorting.
-    resources.sort(Comparator.comparing(r -> r.getIdElement().getIdPart()));
+    var entitiesWithType =
+        resourceTypes.stream()
+            // There may be multiple resource types which will each result in a list of claim
+            // entities. So, to ensure that we have a flat list of entities to their type, we use
+            // flatMap
+            .flatMap(
+                type ->
+                    claimDao
+                        .findAllByMbiAttribute(
+                            type, mbi, bundleOptions.isHashed, lastUpdated, serviceDate)
+                        .stream()
+                        .map(e -> new ImmutablePair<>(e, type)))
```

Comment:
I don't feel _too_ strongly about this, but considering the degree of nesting here, maybe this is worth breaking into a helper function?

Reply:
I'm hesitant about refactoring this, but I definitely could be convinced otherwise. Thinking about it, the resulting helper method would have a rather complicated argument list (all arguments to `findAllByMbiAttribute`) and return type (`Stream<ImmutablePair<ResourceTypeV2...`). Additionally, although this is far less relevant, I can't think of any other spot in the code where this hypothetical helper method would be useful other than to replace lines 495 - 499 in this file.

I feel like this would result in more code for not too much gain. I'm no fan of unnecessary nesting, but I feel this is readable enough and anyone with a good grasp of `Stream`s should be able to grok the intent fairly quickly (take a list of `ResourceTypeV2`s, for each find list of claim entities of that type, flat map each type to each entity in the aforementioned list by pairing them using `ImmutablePair`).